[make:decorator] Add new maker to create decorator

WedgeSama opened the pull request:
Add new maker to create decorator.
https://github.com/symfony/maker-bundle/issues/1401

Allow to create new decorator for existing services.

It try first to decorate by implements:
- The service ID used is an interface => implements this interface
- The service class implements one (or more) interface(s) and it do not declare more public method => implements those interfaces

If cannot implements, it will fallback to extends:
- Simply extends the service's class
- Do not work for final class

e.g.

```shell
bin/console make:decorator Symfony\Component\Routing\Generator\UrlGeneratorInterface MyUrlGenerator
```

```php
<?php

namespace App;

use Symfony\Component\DependencyInjection\Attribute\AsDecorator;
use Symfony\Component\DependencyInjection\Attribute\AutowireDecorated;
use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
use Symfony\Component\Routing\RequestContext;

#[AsDecorator(UrlGeneratorInterface::class)]
final class MyUrlGenerator implements UrlGeneratorInterface
{
    public function __construct(
        #[AutowireDecorated]
        private readonly UrlGeneratorInterface $inner,
    ) {
    }

    public function generate(string $name, array $parameters = [], int $referenceType = self::ABSOLUTE_PATH): string
    {
        return $this->inner->generate($name, $parameters, $referenceType);
    }

    public function setContext(RequestContext $context): void
    {
        $this->inner->setContext($context);
    }

    public function getContext(): RequestContext
    {
        return $this->inner->getContext();
    }
}
```

WedgeSama committed:
[make:decorator] Add new maker to create decorator

https://github.com/symfony/maker-bundle/issues/1401

Neirda24 commented on src/Maker/MakeDecorator.php:
```diff
@@ -0,0 +1,141 @@
+
+            ($question = new Question($argument->getDescription()))
+                ->setAutocompleterValues($this->ids)
+                ->setValidator(fn ($answer) => Validator::serviceExists($answer, $this->ids))
```
based on the issue I'd suggest that the command accept not only the id but a classname and tries to guess the id based on it (if the service does not exists for example)

Neirda24 replied:
3 yes with fault tolerance maybe ? `serializer` vs `serialiser` There are example in symfony's code about this.

WedgeSama replied:
Done. I also did guess from class/interface short name.

Jean-Beru commented on tests/DependencyInjection/Fixtures/ServiceA.php:
```diff
@@ -0,0 +1,25 @@
+        return 'service_a';
+    }
+
+    public function getDefault(string $mode = self::MODE_FOO): ?string
```
Maybe we can use this test to check different argument representations. Ex: without type, without default value, PHP8 attributes, variadic arguments, etc.)

WedgeSama replied:
I test it here [tests/Util/ClassSource/ClassMethodTest.php#L62](https://github.com/WedgeSama/maker-bundle/blob/feature/make-decorator/tests/Util/ClassSource/ClassMethodTest.php#L62)
Is it enough?

WedgeSama committed:
Fix argument default value.

PierreCapel commented on src/Maker/MakeDecorator.php:
```diff
@@ -0,0 +1,141 @@
+
+    public static function getCommandDescription(): string
+    {
+        return 'Create CRUD for Doctrine entity class';
```
```suggestion
        return 'Create a decorator of a service class';
```

WedgeSama replied:
Done

WedgeSama committed:
Add DecoratorHelper + Suggest service's id on error/shortname + priority/onInvalid

Also fix:
- ignore `__construct`
- fix static method on decorate by implements
- Fix description

WedgeSama commented on src/Maker/MakeDecorator.php:
```diff
@@ -0,0 +1,141 @@
+        $command
+            ->addArgument('id', InputArgument::OPTIONAL, 'The ID of the service to decorate.')
+            ->addArgument('decorator-class', InputArgument::OPTIONAL, \sprintf('The class name of the service to create (e.g. <fg=yellow>%sDecorator</>)', Str::asClassName(Str::getRandomTerm())))
+            ->setHelp($this->getHelpFileContents('MakeDecorator.txt'))
```
Added

Neirda24 commented on src/Util/DecoratorInfo.php:
```diff
@@ -0,0 +1,270 @@
+                } elseif ($parameter->isDefaultValueAvailable()) {
+                    $defaultValue = $parameter->getDefaultValue();
+
+                    if (\is_string($defaultValue)) {
```
is variadic also supported ?

Neirda24 commented on src/Util/DecoratorInfo.php:
```diff
@@ -0,0 +1,270 @@
+            return implode('&', array_map($this->parseType(...), $type->getTypes()));
+        }
+
+        throw new RuntimeCommandException('Should never be reach.');
```
```suggestion
        throw new RuntimeCommandException('Should never be reached.');
```

WedgeSama commented:
Sorry, I was in holiday :sweat_smile: 
No problem @GromNaN  :wink:
